### PR TITLE
derivation for isomorphic things (#1662)

### PR DESCRIPTION
**Pull request by valydia**
Following up on #1720

### CODE REVIEW

File: core/src/main/scala/scalaz/Isomorphism.scala [hunk 1]
```diff
@@ -609,22 +607,21 @@ trait IsomorphismArrow[F[_, _], G[_, _]] extends Arrow[F] with IsomorphismSplit[
 
 
 trait IsomorphismBifoldable[F[_, _], G[_, _]] extends Bifoldable[F] {
-  protected[this] def biNaturalTrans: F ~~> G
+  def iso: F <~~> G
```
Comment (xuwei-k) [inline core/src/main/scala/scalaz/Isomorphism.scala]: https://travis-ci.org/scalaz/scalaz/jobs/376128922#L945-L958

```

[info] scalaz-core: found 6 potential binary incompatibilities while checking against org.scalaz:scalaz-core_2.11:7.2.22 
[error]  * method naturalTrans()scalaz.NaturalTransformation in trait scalaz.IsomorphismTraverse does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scalaz.IsomorphismTraverse.naturalTrans")
[error]  * method biNaturalTrans()scalaz.BiNaturalTransformation in trait scalaz.IsomorphismBifoldable does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scalaz.IsomorphismBifoldable.biNaturalTrans")
[error]  * method iso()scalaz.Isomorphisms#Iso3 in trait scalaz.IsomorphismBifoldable is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("scalaz.IsomorphismBifoldable.iso")
[error]  * method naturalTrans()scalaz.NaturalTransformation in trait scalaz.IsomorphismFoldable does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scalaz.IsomorphismFoldable.naturalTrans")
[error]  * method iso()scalaz.Isomorphisms#Iso2 in trait scalaz.IsomorphismFoldable is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("scalaz.IsomorphismFoldable.iso")
[error]  * method biNaturalTrans()scalaz.BiNaturalTransformation in trait scalaz.IsomorphismBitraverse does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scalaz.IsomorphismBitraverse.biNaturalTrans")
java.lang.RuntimeException: scalaz-core: Binary compatibility check failed!
```

Comment (valydia) [reply core/src/main/scala/scalaz/Isomorphism.scala]: @emilypi I received some comments, that don't seem to appear on the page, about fixing previous implementations of `Isomorphism` for `Foldable` and `Bifoldable`, is my second commit what you had in mind! Obviously, what I've done breaks binary compatibility...

**Comment by xuwei-k** [comment]
I don’t like so much boilerplate code 😕 
Should we auto generate like this? https://github.com/xuwei-k/scalaz/commit/8616790b68840772058db0b682af99e5bd974962 WDYT?

**Comment by xuwei-k** [comment]
https://github.com/scalaz/scalaz/pull/1842

**Pull request closed without merging**